# 搜集用户信息与数据整合

* 搜集用户发帖comment id
* 如何解码发帖comment id？
    * 正向编码方法
    * 如何逆向破解？
    
* 爬取用户信息
* 数据整合

## 1. 整理发帖人信息

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
address = '../../../Danmu_files/'

In [12]:
files = os.listdir(address)
files

['.DS_Store',
 'Senti2_B站剧集.mov',
 'Senti3_SeasonID_EpisodeID.mov',
 'Senti4_取当前弹幕.mov',
 'Senti5_取历史弹幕.mov',
 'Senti6_TFIDF.mov',
 'Senti7_弹幕频率_Sentimental.mov',
 'Spam',
 'UID_UCMID.csv',
 '宣传图片(请将视频制作为gif动图_谢谢啦)',
 '番外篇第1集.csv',
 '番外篇第2集.csv',
 '番外篇第3集.csv',
 '番外篇第4集.csv',
 '番外篇第5集.csv',
 '第一季第10集.csv',
 '第一季第11集.csv',
 '第一季第12集.csv',
 '第一季第1集.csv',
 '第一季第2集.csv',
 '第一季第3集.csv',
 '第一季第4集.csv',
 '第一季第5集.csv',
 '第一季第6集.csv',
 '第一季第7集.csv',
 '第一季第8集.csv',
 '第一季第9集.csv',
 '第三季第10集.csv',
 '第三季第11集.csv',
 '第三季第12集.csv',
 '第三季第1集.csv',
 '第三季第2集.csv',
 '第三季第3集.csv',
 '第三季第4集.csv',
 '第三季第5集.csv',
 '第三季第6集.csv',
 '第三季第7集.csv',
 '第三季第8集.csv',
 '第三季第9集.csv',
 '第二季第10集.csv',
 '第二季第11集.csv',
 '第二季第12集.csv',
 '第二季第1集.csv',
 '第二季第2集.csv',
 '第二季第3集.csv',
 '第二季第4集.csv',
 '第二季第5集.csv',
 '第二季第6集.csv',
 '第二季第7集.csv',
 '第二季第8集.csv',
 '第二季第9集.csv',
 '第四季第10集.csv',
 '第四季第11集.csv',
 '第四季第1集.csv',
 '第四季第2集.csv',
 '第四季第3集.csv',
 '第四季第4集.csv',
 '第四季第5集.csv',
 '第四季第6集.csv',
 '第四季第7集.csv',
 '第四季第8集.csv',
 '第四季第9

comment_ids = set()
for file in files[10:-1]:
    tmp=pd.read_csv(address+file)
    file_id_set = set(tmp['ID'])
    comment_ids = comment_ids|file_id_set
    print(file,len(comment_ids))

In [ ]:
comment_ids = set() #set() element must be unique.
for file in files[10:-1]: 
    tmp=pd.read_csv(address+file) 
    file_id_set = set(tmp['ID']) #extract the comment ID from danmu csv files from Each episode from Each season, 
                                 # and store them into a set, then convert to pd-df, then convert to csv called "Comments_ids.csv" (see below)
                                # note: each danmu csv files from each episode from each season is not provided, it is obtained from webcrawler,
                                # but comment_ids.csv was provided in the current folder.
    comment_ids = comment_ids|file_id_set 
    print(file,len(comment_ids))

* 查看所有comment_ids

In [26]:
comment_ids # this is set

{'1f7dd51a',
 'cc108eca',
 '83eff729',
 '41870812',
 'fb64cc6d',
 'D4281098',
 'b50bdd33',
 '5c707a92',
 'e045cbae',
 'a0db3559',
 '6f089207',
 '1f8d5719',
 '897ead8f',
 '62fb5261',
 '8844e873',
 '74cc309a',
 '812f616f',
 'f523d0ff',
 '7eca2422',
 'bfb44445',
 '68a2ed93',
 'd56f84c8',
 '32504f6f',
 '2cea4a1d',
 'D2d914da',
 'ed956b71',
 '35d62bd5',
 'f83c11f6',
 '63060e70',
 'f112ee40',
 'a55ee8c9',
 'D4bf324b',
 '272bad7d',
 '60d47480',
 'bc3459fd',
 '7fbe45c9',
 'a9d8458d',
 'a57f0db8',
 '5a5ebec9',
 'c29749c9',
 'e4240141',
 'f00a21b9',
 '80872a88',
 '438ad843',
 '4de42bd9',
 '7091b20a',
 'd678b394',
 '6c4b0d11',
 '3721ccc8',
 'Db0a5178',
 '850b4588',
 '129349f0',
 'dc3db13d',
 '0feb8257',
 'Df64d853',
 '30b46fe9',
 '4118068a',
 '645dad2a',
 'a13172c3',
 '9bb72b59',
 '63c3ef94',
 'e26fadb2',
 'b0f1d6f2',
 '994c48e9',
 'e15b744f',
 '23af51c1',
 '12e9b62c',
 '45e84e96',
 '146a1318',
 '6ea0d804',
 '967519d2',
 'Da856912',
 'ca764643',
 '73e77c14',
 'c28ab4a6',
 '04d1e071',
 '44cd567e',

* 转化为dataframe
    * ids_pd = pd.DataFrame(list(comment_ids),columns=['ID'])

In [10]:
ids_pd = pd.DataFrame(list(comment_ids),columns=['ID'])

* 及时保存：ids_pd.to_csv('comments_ids.csv')

* 下次读取 
    * comment_ids = pd.read_csv('comments_ids.csv')
    * comment_ids.drop(['Unnamed: 0'],inplace=True,axis=1)

In [9]:
comment_ids = pd.read_csv('comments_ids.csv')

In [10]:
comment_ids.shape

(365891, 2)

In [4]:
comment_ids.drop(['Unnamed: 0'],inplace=True,axis=1)

In [5]:
ids_pd=comment_ids
comment_ids=set(comment_ids['ID'])


In [6]:
comment_ids

{'78c46605',
 'd45da2c2',
 '00de9bf6',
 '3bff04ef',
 'c4e22f4b',
 'd6daa9ab',
 '775fc815',
 'De3efe59',
 'b75e0e35',
 '397f5f33',
 'b3ebddba',
 '5be1caba',
 '159c8498',
 'b4863741',
 '7b3bfae1',
 '9538a7af',
 '4c812d64',
 'c0f15104',
 'ae2d0139',
 'a92cf162',
 '2604b787',
 '8cdfa6a3',
 '584b380b',
 '8d66f28e',
 'D86e953b',
 'c67d4ee0',
 'efb9c9c0',
 '223d4901',
 'c17d1572',
 'f8151373',
 '9c542d03',
 'ecc0bd15',
 'fdac8dce',
 '2312b2e4',
 '07b3e83f',
 'D69c85a6',
 'ca5e3d06',
 '2e3a5df8',
 '4abbbf20',
 'e85411b9',
 'aba1cc7c',
 '922ca78c',
 'a55a7af2',
 '0b0c7ae0',
 '01a540d7',
 '3f29e575',
 'D50cb21a',
 'df8a06db',
 '515fd22c',
 'D15c2f8b',
 '3ff0d72f',
 '4332529f',
 'f5ffc766',
 'aa06b165',
 '7e0160e7',
 'bc91b028',
 '72e80998',
 'D178a90f',
 '9f4e2653',
 '04e3b436',
 '29cc8e04',
 '5380d112',
 '5c80b8af',
 'e85b2e4f',
 '620f196a',
 'c2fbdf36',
 'cc26e34c',
 '751d6c02',
 'c6fd42b4',
 'f5c56ce9',
 'ab173e2b',
 'Dc57aae0',
 '1f82f726',
 '117eb8a0',
 '48abebb9',
 'D6f3bd1b',
 'D35e158f',

In [ ]:
# words.lower()

In [75]:
# comment_ids_list = list(comment_ids)
# comment_ids_D=[]
# for i in comment_ids_list:
#     if i.startswith('D'):
#         comment_ids_D.append(i.lower())

* 将所有ID小写

In [7]:
comment_ids_Lower_list=[]
comment_ids_list = list(comment_ids)
for i in comment_ids_list:
    comment_ids_Lower_list.append(i.lower())

comment_ids_Lower_set = set(comment_ids_Lower_list)

In [8]:
comment_ids_Lower_set

{'d697eb6b',
 '78c46605',
 'd45da2c2',
 '00de9bf6',
 '3bff04ef',
 'c4e22f4b',
 'd6daa9ab',
 '775fc815',
 'b75e0e35',
 '397f5f33',
 'b3ebddba',
 '5be1caba',
 '159c8498',
 'b4863741',
 '7b3bfae1',
 '9538a7af',
 '4c812d64',
 'c0f15104',
 'ae2d0139',
 'a92cf162',
 '2604b787',
 '8cdfa6a3',
 '584b380b',
 '8d66f28e',
 'c67d4ee0',
 'efb9c9c0',
 '223d4901',
 'c17d1572',
 'f8151373',
 '9c542d03',
 'ecc0bd15',
 'fdac8dce',
 '2312b2e4',
 '07b3e83f',
 'dc5196b6',
 'ca5e3d06',
 '2e3a5df8',
 '4abbbf20',
 'e85411b9',
 'd1eec3eb',
 'aba1cc7c',
 '922ca78c',
 'a55a7af2',
 '0b0c7ae0',
 '01a540d7',
 '3f29e575',
 'ddfa179c',
 'df8a06db',
 '515fd22c',
 '3ff0d72f',
 '4332529f',
 'f5ffc766',
 'dbc52182',
 'aa06b165',
 'd42371af',
 '7e0160e7',
 'bc91b028',
 'db53f6b7',
 '72e80998',
 'da219f1c',
 '9f4e2653',
 '04e3b436',
 'd64e133c',
 'd7dffe79',
 '29cc8e04',
 '5380d112',
 '5c80b8af',
 'e85b2e4f',
 '620f196a',
 'c2fbdf36',
 'cc26e34c',
 '751d6c02',
 'c6fd42b4',
 'f5c56ce9',
 'ddcf340d',
 'ab173e2b',
 '1f82f726',

In [4]:
# comment_ids_D_lower_set = set(comment_ids_D)

In [2]:
# 'Dd132'.startswith('D')

In [3]:
# comment_ids_D_lower_set

In [68]:
comment_ids_Lower_set

{'01cd844c',
 '888e0716',
 '9a966b79',
 '279df27a',
 'd9f554a9',
 'db2bc2a3',
 '866853f7',
 'dbcbcad7',
 '87ce17a6',
 '4e3c71de',
 'd3028d5e',
 '6b0b8d85',
 'a4e8d403',
 '253e4e34',
 'b5bb62f7',
 '508a8cde',
 '8ce7a99b',
 '64cc7fc9',
 '88415704',
 'b70fb365',
 '8170ed31',
 '9a12b5f2',
 '9dc6abf7',
 'ff426c44',
 'cc5eaa75',
 '6cd13b2b',
 '483ed0e8',
 '454cc54e',
 '2053dffe',
 'd8e9414c',
 'd045b593',
 'cf3151bd',
 '88acf36e',
 'd1d8a04f',
 'e23dbdf9',
 'f1bbc2ad',
 '7aa3bbc7',
 'c1031258',
 '37f5b59b',
 '1ee7b0d0',
 '2a28fde2',
 '7440c618',
 'aa582478',
 'd41d9f2d',
 'e64da788',
 '26745041',
 '485e7e6f',
 '8b8602bb',
 '2a91532e',
 'e16e758a',
 '58f934aa',
 'a6505c1a',
 '4686c822',
 'd70fdbf8',
 'd36c4c61',
 '395e2ccb',
 'bed87708',
 'd8534664',
 '18100e4e',
 '6a2ddf01',
 'e3aff7a8',
 'de6715a7',
 'd020abca',
 'df043b1b',
 '13c83e7f',
 'd8f8d8d7',
 'bc77c17b',
 'b90165fe',
 '9e091f37',
 '56ffa758',
 '16b79fa0',
 '6d5249a2',
 '603d77f8',
 '7cb1376e',
 'd0385992',
 'fc4ac565',
 'd7711026',

## 2. 如何解码发帖id?


In [ ]:
mid = 1 - 2亿4千万 #member ID is pure number. 

### 2.1 正向编码方法：zlib.crc32


In [ ]:
# example of comment id: 8ce7a99b

In [ ]:
 mid  -> comment id   # bilibili uses zlib.crc32 to encode mid to comment id.

In [9]:
import zlib
def crc32(bdata):
    h = zlib.crc32(bdata)
    x='%08X' % (h & 0xffffffff,)
    return(x.lower())

In [10]:
crc32(b'34566663')

'8bf01fb2'

In [11]:
crc32(b'174918342')

'd31cf5b9'

In [11]:
# h = zlib.crc32(b'34566663')
# '%08X' % (h & 0xffffffff,)

In [12]:
all_crcs = pd.DataFrame(index=range(5),columns=['ID'])
for i in range(5):
    tmp = bytes(str(i).encode('utf8'))
    all_crcs.loc[i]['ID'] = crc32(tmp)

In [13]:
all_crcs

ID
0  f4dbdf21
1  83dcefb7
2  1ad5be0d
3  6dd28e9b
4  f3b61b38

### 2.2 如何逆向解码？

We have comment id, our purpose is to get mid using [decoding of the comment id].

In [ ]:
mid                              ->     comment id    #crc32
mid [0,2 4000 0000]                      crc32(mid)

In [ ]:
想要的是mid， 手里面有 comment id

* 有网站可以逆向解码，是否使用？
* danmu.fuckbilibili.com
* fuckbilibili.com/bilidanmaku.html

- Method 1: use website above to decode for us:
    - drawback: slow, decode each comment id needs 2s, comment id has 365891 rows. If we use this method, it needs 9 days

* Method 2: 
    - Step 1. Convert mid from 0 -> 240,000,000 to comment id (dictionary), 
    - Step 2. then find each comment id from 365891 rows in dictionary (240,000,000 comment id dictionary)  -- see below, it is quicker than method 1.

We decided to use Method 2

Step 1.

In [14]:
all_crcs = pd.DataFrame(index=range(240000000),columns=['ID'])
for i in range(5):
    tmp = bytes(str(i).encode('utf8'))
    all_crcs.loc[i]['ID'] = crc32(tmp)

In [16]:
all_crcs.to_csv('UID_UCMID.csv') # write 240,000,000 mid to comment id --> store into a csv file, size = 4 G

In [8]:
test1 = pd.read_csv('UID_UCMID.csv')

In [10]:
test1.head(20)

Unnamed: 0     ucmid
0            0  f4dbdf21
1            1  83dcefb7
2            2  1ad5be0d
3            3  6dd28e9b
4            4  f3b61b38
5            5  84b12bae
6            6  1db87a14
7            7  6abf4a82
8            8  fa005713
9            9  8d076785
10          10  a15d25e1
11          11  d65a1577
12          12  4f5344cd
13          13  3854745b
14          14  a630e1f8
15          15  d137d16e
16          16  483e80d4
17          17  3f39b042
18          18  af86add3
19          19  d8819d45

In [14]:
test1.shape

(240000000, 2)

Step 2. 用户ID一一对比

In [10]:
decode_file_name = './UID_UCMID.csv'

In [ ]:
ids_pd['orig_ID']=pd.Series(dtype=int)

In [18]:
# 注意对比顺序:
    # Order 1: 
        # For each row in 4 G file (240 million rows):
            # for each row in comment ids (360K rows)
                #match comment ids and mid
    # Order 2:
        # for each row in comment ids (360K rows):
            # for each row in 4 G file (240million rows)
                #match comment ids and mid
                
    # The two orders have the same complexity. But in practice, Order 1 takes 8 hours while order 2 takes 5 days, why?
        # - Because computer will put outer loop table into a RAM, then refer to it directly rather than read from hard drive. 
        # In order 1, 4G file has already been in RAM when first time loop, so later on, computer only needs to read 360K rows
        # inner loop table from hard drive; While in order 2, only 360K was read into RAM, computer needs to read 4G file from 
        # hard drive everytime when it goes through the innerloop.
        # So order 1 is faster than order 2 in this case, we choose order 1, see below
import re
with open(decode_file_name,'r') as f:
    for i,line in enumerate(f): # for each row in that 4G file (240Million rows)
        if (i-1)%10000==0: #for every 10000 rows, print something to give user a Feedback, telling you computer is still running
            print(i-1)
        if (i-1)%500000==0:
            gc.collect()
        if (i-1)%5000000==0: #for every 5 million rows, store the partial comparison results to csv, just in case computer is power outage.
            ids_pd.to_csv('ids_pd%d.csv'%(int((i-1)//1000000)))
        match=re.search(r",([a-zA-Z0-9]+)",line) # IF line matches the Regex format, it is one user comment id and can be used to search in comment_ids 
        # this step can be used to avoid selecting e.g. NaN, empty item in ucmid
        
        user_cmid = match.group(1)
        if user_cmid in comment_ids: #search if user comment id in comment_ids set
            ids_pd.loc[ids_pd['ID']==user_cmid,'orig_ID']=i-1

In [6]:
ids_pd = pd.read_csv('ids_pd235.csv') # ids_pd235.csv - this is one of the partial comparison results from above, for demo.
ids_pd.head(6) # ID = comment id. orig_ID = matched mid.

Unnamed: 0        ID      orig_ID
0           0  1f7dd51a    8976061.0
1           1  cc108eca   87371356.0
2           2  83eff729    8961841.0
3           3  41870812  210120754.0
4           4  fb64cc6d    6802440.0
5           5  D4281098          NaN

* As we can see, some comment ID is not encoded using crc32, so we cannot find its original mid. (e.g. comment ID Started with "D", we cannot find its original mid.
* By scanning, we found only 2% data has NaN. So we can drop those data using formula below
* 去掉多余列：ids_pd.drop(columns=['Unnamed: 0'],axis=1,inplace=True)

In [60]:
# ids_pd_lower_d_list = []
# for i in ids_pd['ID']:
#     ids_pd_lower_d_list.append(i.lower())

# ids_pd_lower_d = pd.DataFrame(ids_pd_lower_d_list,columns=['ID'])
# ids_pd_lower_d['orig_ID'] = ids_pd['orig_ID']

In [44]:
# tmp_id = 'D4281098'
# if tmp_id in comment_ids_D_set:
#     print('yes')

## 3. 现在有了用户的ID, 采集用户信息
* 整理用户mid ID
* 爬取用户信息
* url = 'http://space.bilibili.com/ajax/member/GetInfo?mid=' + str(int(i))
* Thanks to airingursb/bilibili-user


### 3.1 整理用户mid ID

In [4]:
import pandas as pd
import numpy as np
ids_pd = pd.read_csv('./ids_pd235.csv') # use ids_pd235.csv as an example to demo

In [5]:
ids = ids_pd['orig_ID']
ids

0           8976061.0
1          87371356.0
2           8961841.0
3         210120754.0
4           6802440.0
5                 NaN
6           5814204.0
7          15348789.0
8          29011305.0
9          11377776.0
10         39400069.0
11         11971103.0
12         17635757.0
13         19898877.0
14         54748598.0
15         96954891.0
16          8324958.0
17         66768332.0
18         12060048.0
19          8048759.0
20           758897.0
21          9025333.0
22         28663067.0
23         60349382.0
24                NaN
25          2352386.0
26          9151091.0
27         22663076.0
28         44110686.0
29         46049735.0
             ...     
365861      5124802.0
365862     27103794.0
365863     18432399.0
365864      1716312.0
365865     11480249.0
365866     96221717.0
365867    153348943.0
365868      8670310.0
365869    215110807.0
365870     28835131.0
365871            NaN
365872      7304214.0
365873     72230841.0
365874     52207969.0
365875    

### 3.2 爬取用户信息

* 准备url序列

In [6]:
urls=[]
for i in ids:
    if not np.isnan(i):
        url = 'http://space.bilibili.com/ajax/member/GetInfo?mid=' + str(int(i))
    urls.append(url)


In [7]:
urls #urls to each userid (mid)

['http://space.bilibili.com/ajax/member/GetInfo?mid=8976061',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=87371356',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=8961841',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=210120754',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=6802440',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=6802440',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=5814204',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=15348789',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=29011305',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=11377776',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=39400069',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=11971103',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=17635757',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=19898877',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=54748598',
 'http://space.bilibili.com/ajax/member/GetInfo?mid=9695489

In [10]:
import requests
import json
import random
import pymysql # there are other module like MySQL-python, which is really hard to install
from multiprocessing.dummy import Pool as ThreadPool
import sys
import datetime
import time
from imp import reload
import pandas as pd
import stem
from stem import Signal
from stem.control import Controller
import numpy as np

* 准备相应函数 to bypass website's 反爬虫function 

In [16]:
def datetime_to_timestamp_in_milliseconds(d):
    def current_milli_time(): return int(round(time.time() * 1000))
    return current_milli_time()

reload(sys)

def LoadUserAgents(uafile):
    """
    uafile : string
        path to text file of user agents, one per line
    """
    uas = []
    with open(uafile, 'rb') as uaf:
        for ua in uaf.readlines():
            if ua:
                uas.append(ua.strip()[1:-1 - 1])
    random.shuffle(uas)
    return uas

uas = LoadUserAgents("user_agents.txt") #伪装不同的浏览器访问
head = {
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36',
    'X-Requested-With': 'XMLHttpRequest',
    'Referer': 'http://space.bilibili.com/45388',
    'Origin': 'http://space.bilibili.com',
    'Host': 'space.bilibili.com',
    'AlexaToolbar-ALX_NS_PH': 'AlexaToolbar/alx-4.0',
    'Accept-Language': 'zh-CN,zh;q=0.8,en;q=0.6,ja;q=0.4',
    'Accept': 'application/json, text/javascript, */*; q=0.01',
}

* 准备DataFrame(placeholder) to store user info

In [17]:
user_info = pd.DataFrame(columns=['mid','name','sex','rank','face','coins','regtime','spacesta','place',\
                'birthday','sign','description','article','level','level_current_exp','level_next_exp',\
                'nameplate_nid','nameplate_name','nameplate_level','nameplate_condition','official_verify_type',\
                'toutu','toutuId','following','fans','playNum']) #26 columns
user_info = user_info.astype({'mid':np.str,'name':np.str,'sex':np.str,'rank':np.int8,'face':np.str,'coins':np.int8,\
                              'regtime':np.int64,'spacesta':np.int16,'place':np.str,\
                'birthday':np.str,'sign':np.str,'description':np.str,'article':np.int8,'level':np.int8,\
                'level_current_exp':np.int8,'level_next_exp':np.int8,\
                'nameplate_nid':np.str,'nameplate_name':np.str,'nameplate_level':np.str,\
                'nameplate_condition':np.str,'official_verify_type':np.str,\
                'toutu':np.str,'toutuId':np.str,'following':np.int8,\
                              'fans':np.int8,'playNum':np.int8})

* 采集信息的主要函数

In [42]:
def getsource(url): #input is an url of one user(mid) -- one mid corresponds to one url -- each url comes from "urls" generated above

    payload = {
        '_': datetime_to_timestamp_in_milliseconds(datetime.datetime.now()),
        'mid': url.replace('http://space.bilibili.com/ajax/member/GetInfo?mid=', '')
    }

    time.sleep(0.6)
    
    ua = random.choice(uas)

    head = {'User-Agent': ua,
            'Referer': 'http://space.bilibili.com/' + str(random.randint(9000, 150000)) + '/'
            } #

    #proxy = {}#random.choice(proxies) #if you have multiple usable IPs,you can store them into the proxy dictionary and use like below:
    # "proxies=proxy" see below. (It will help you bypass more anti-web-crawler function in website)
    
    jscontent = requests.Session().post('http://space.bilibili.com/ajax/member/GetInfo',
                headers=head,data=payload,timeout=10).text         #timeout=10 means: 10s no response from website, jump to next url          
                # ,proxies=proxy).text        
    
    row = list(np.zeros(26)) #placeholder, 26 colums corresponding to user_info
    
    try:
        jsDict = json.loads(jscontent) # jscontent obtained from website above is a dictionary format object.
        statusJson = jsDict['status'] if 'status' in jsDict.keys() else False
        if statusJson == True:
            if 'data' in jsDict.keys():
                jsData = jsDict['data']
                mid = jsData['mid']
                name = jsData['name']
                sex = jsData['sex']
                rank = jsData['rank']
                face = jsData['face']
                coins = jsData['coins']
                if 'regtime' in jsData.keys(): #registry time
                    regtime = jsData['regtime'] 
                else:
                    regtime = 0
                spacesta = jsData['spacesta']
                place = jsData['place'] if 'place' in jsData.keys(
                ) else 'noplace'
                birthday = jsData['birthday'] if 'birthday' in jsData.keys(
                ) else 'nobirthday'
                sign= jsData['sign']
                description = jsData['description']
                article = jsData['article']
                level = jsData['level_info']['current_level']
                level_current_exp = jsData['level_info']['current_exp']
                level_next_exp = jsData['level_info']['next_exp']
                nameplate_nid = jsData['nameplate']['nid']
                nameplate_name = jsData['nameplate']['name']
                nameplate_level = jsData['nameplate']['level']
                nameplate_condition = jsData['nameplate']['condition']
                official_verify_type = jsData['official_verify']['type']
                toutu = jsData['toutu']
                toutuId = jsData['toutuId']
                playNum = jsData['playNum']
                exp = jsData['level_info']['current_exp']
                try:
                    res = requests.get(
                    'http://api.bilibili.com/x/relation/stat?vmid=' + str(mid) + '&jsonp=jsonp').text
                    js_fans_data = json.loads(res)
                    following = js_fans_data['data']['following']
                    fans = js_fans_data['data']['follower']
                except:
                    following = 0
                    fans = 0
                    
                row =  [mid,name,sex,rank,face,coins,regtime,spacesta,place,\
                birthday,sign,description,article,level,level_current_exp,level_next_exp,\
                nameplate_nid,nameplate_name,nameplate_level,nameplate_condition,\
                official_verify_type,toutu,toutuId,following,fans,playNum]
                
                print('get id=%s, name=%s'% (mid,name))

            else:
                print('no data now')
                

        else:
            print('Error in JsonStatus')
            #print("Error: " + url)
            #Error_JSON_url.append(url)

    except ValueError: #IP被反爬虫封了
        print('Perhaps been found and rejected, retriving has failed')
        time.sleep(100)
        #Error_reject_url.append(url)
        #renew_connection()
    
    
    
    return row # return the info (26 columns) for this user(url)

--------------An example for extracting user info of userid 758897------------

In [19]:
i =(ids[20])
url = 'http://space.bilibili.com/ajax/member/GetInfo?mid=' + str(int(i))
url

'http://space.bilibili.com/ajax/member/GetInfo?mid=758897'

In [20]:
payload = {
    '_': datetime_to_timestamp_in_milliseconds(datetime.datetime.now()),
    'mid': url.replace('http://space.bilibili.com/ajax/member/GetInfo?mid=', '')
}

ua = random.choice(uas)

head = {'User-Agent': ua,
        'Referer': 'http://space.bilibili.com/' + str(random.randint(9000, 10000)) + '/'
        } #反爬虫时用的

#proxy = {}#random.choice(proxies)

jscontent = requests.Session().post('http://space.bilibili.com/ajax/member/GetInfo', headers=head,
            data=payload).text                   
            #headers=head, data=payload, proxies=proxies).text

In [21]:
jscontent

'{"status":true,"data":{"mid":"758897","name":"starfer","approve":false,"sex":"\\u7537","rank":"10000","face":"http:\\/\\/i0.hdslb.com\\/bfs\\/face\\/d12d6aaaa0607b782a38ece0f1c2fa826f842fbe.jpg","coins":0,"DisplayRank":"10000","regtime":1360390285,"spacesta":0,"place":"\\u5c71\\u897f\\u7701 \\u6714\\u5dde\\u5e02","birthday":"0000-01-13","sign":"","description":"","article":0,"level_info":{"next_exp":28800,"current_level":5,"current_min":10800,"current_exp":13766},"pendant":{"pid":107,"name":"\\u5168\\u804c\\u9ad8\\u624b","image":"http:\\/\\/i2.hdslb.com\\/bfs\\/face\\/4fade9d625b7b2b8cadbd93106c1f10457e131c3.png","expire":1546444800},"nameplate":{"nid":60,"name":"\\u2028\\u996d\\u5708\\u840c\\u65b0","image":"http:\\/\\/i1.hdslb.com\\/bfs\\/face\\/51ca16136e570938450bca360f28761ceb609f33.png","image_small":"http:\\/\\/i2.hdslb.com\\/bfs\\/face\\/9abfa4769357f85937782c2dbc40fafda4f57217.png","level":"\\u666e\\u901a\\u52cb\\u7ae0","condition":"\\u5f53\\u524d\\u6301\\u6709\\u7c89\\u4e1d\\

-----------------end of example----------------------

Let's try to extract part of user ids' user info for demo:

- Store extracted user info to csv

In [43]:
rows = []
i = 0
for url in urls[2200:2600]: #extract ids from 2200th to 2600th for demo and store into one csv for every 300 ids
    
    if i%100==99:
        time.sleep(50)
    
    try:
        row = getsource(url) #use getsource function defined above to extract user info and stored into "row"
    except requests.exceptions.ConnectTimeout:
        row[0] = url
    except requests.exceptions.Timeout:
        row[0] = url
    
    rows.append(row)

    if i%300==0:
        tmp_pd = pd.DataFrame(data=rows,columns=['mid','name','sex','rank','face','coins','regtime','spacesta','place',\
                'birthday','sign','description','article','level','level_current_exp','level_next_exp',\
                'nameplate_nid','nameplate_name','nameplate_level','nameplate_condition','official_verify_type',\
                'toutu','toutuId','following','fans','playNum'])
        user_info = user_info.append(tmp_pd,ignore_index=True)
        user_info.to_csv('user_%d.csv'%i)
        
    i = i+1
    
    


Perhaps been found and rejected, retriving has failed


KeyboardInterrupt: 

In [39]:
user_info

mid            name sex   rank  \
0    8976061      半生浮尘曲今世解语花  保密  10000   
1   87371356       Mima酱2333   女  10000   
2    8961841            八条鲶鱼  保密  10000   
3  210120754  d1cqyj93b5mgs2  保密   5000   
4    6802440             南极冬      10000   

                                                face  coins     regtime  \
0  http://i2.hdslb.com/bfs/face/d214e0c109846ad13...      0  1427206959   
1  http://i1.hdslb.com/bfs/face/351b0e0eda4793bd0...      0  1486024136   
2  http://i0.hdslb.com/bfs/face/6be7b0edaffa2d37d...      0  1427122344   
3   http://static.hdslb.com/images/member/noface.gif      0  1505014682   
4  http://i0.hdslb.com/bfs/face/c8d4e8b90de0d4fdc...      0           0   

   spacesta    place    birthday   ...    nameplate_nid nameplate_name  \
0         2           0000-11-25   ...                0                  
1         0           0000-03-15   ...                0                  
2         2           0000-01-01   ...                0                  
3       -10           0000-01-01   ...                0                  
4         0  noplace  nobirthday   ...                6          高级搬运工   

   nameplate_level  nameplate_condition  official_verify_type  \
0                                                          -1   
1                                                          -1   
2                                                          -1   
3                                                          -1   
4             高级勋章      转载视频投稿通过总数>=100                    -1   

                                               toutu  toutuId following  \
0  bfs/space/768cc4fd97618cf589d23c2711a1d1a729f4...        1        43   
1  bfs/space/768cc4fd97618cf589d23c2711a1d1a729f4...        1        59   
2  bfs/space/768cc4fd97618cf589d23c2711a1d1a729f4...        1        98   
3  bfs/space/768cc4fd97618cf589d23c2711a1d1a729f4...        1         0   
4  bfs/space/373e127e8784d3e4c1b5e6db0c27702ba077...       19        75   

    fans  playNum  
0      0        0  
1      0        0  
2      0        0  
3      0        0  
4  11244  2576292  

[5 rows x 26 columns]

The notes below this line were skipped during the lecture

In [86]:
# pool = ThreadPool(processes=2)
# try:
#     #time3=time.time()
#     results = pool.map(getsource, urls)
#     #time4=time.time()
#     #delta_t = max(60- (time4-time3),0)
#     # 
#     #time.sleep(delta_t)
# except Exception:
#     print('ConnectionError')
#     pool.close()
#     pool.join()


get id=2806745, name=寰宇莜
get id=8976061, name=半生浮尘曲今世解语花
get id=18363533, name=神木骁
get id=87371356, name=Mima酱2333
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found and rejected, retriving has failed
Perhaps been found an

KeyboardInterrupt: 

In [ ]:
# if __name__ == '__main__':

#     time1 = time.time()
#     urls = []
#     for i in ids:
#         url = 'http://space.bilibili.com/ajax/member/GetInfo?mid=' + str(int(i))
#     urls.append(url)
    
# #         renew_connection()
#     time.sleep(random.random()*2)
#     print(datetime.datetime.now())
#     #stop_time = random.randint(1,100)

#     pool = ThreadPool(processes=4)
#     try:
#         #time3=time.time()
#         results = pool.map(getsource, urls)
#         #time4=time.time()
#         #delta_t = max(60- (time4-time3),0)
#         # 
#         #time.sleep(delta_t)
#     except Exception:
#         print('ConnectionError')
#         pool.close()
#         pool.join()
#         continue
#         #time.sleep(20)
#         #pool = ThreadPool(processes=7)
#         #time5=time.time()
#         #results = pool.map(getsource, urls)
#         #time6=time.time()
#         #delta_t = max(60 - (time6-time5),0)
#         #time.sleep(delta_t)

#     time2 = time.time()
#     print('Target is 10000 *20000,right now m =',m,' *20000','use time',str(time2 - time1))

#     pool.close()
#     pool.join()

#     Error_MYSQL_df = pd.DataFrame(Error_MYSQL_url)
#     Error_JSON_df = pd.DataFrame(Error_JSON_url)
#     Error_reject_df = pd.DataFrame(Error_reject_url)

#     Error_MYSQL_df.to_csv('Error_MYSQL_url.csv',mode='a',header=False)
#     Error_JSON_df.to_csv('Error_JSON_url.csv',mode='a',header=False)
#     Error_reject_df.to_csv('Error_reject_url.csv',header=False)

